swarm/storage/dbstore.go

```diff
@@ -75,13 +76,25 @@ type DbStore struct {
 	hashfunc SwarmHasher
 
 	lock sync.Mutex
+
+	// Functions putDataFunc and getFunc are used for
+	// saving and retreiving chunk data from the store.
+	// They must be set on DbStore initialization and must not be nil.
+	// They are used to bypass the default functionality of DbStore with
+	// mock.NodeStorer for testing purposes.
+	putDataFunc func(batch *leveldb.Batch, _ Key, data []byte)
```

Comment:
i wonder why we need this extra layer of abstraction here.
I think instead of `putDataFunc` and `getDataFunc` 
we should push it down to `decodeData` and `encodeData` no?
the mock store interfaces can stay they are nice.

Reply:
The reason for extra layer of abstraction is to avoid checking if the mock store is provided to the DbStore on every Put and Get function call. This way, it is done in the constructor, as whether mock store is used or not is only specified on construction, not to be changed later. If additional *if* in Put and Get (indirectly by decodeData and encodeData) is not a problem, we can loose the abstraction. On the other hand, decodeData and encodeData are simple and very clear on the purpose and functionality.

If I understood correctly, we should add to encodeData the decision whether to save data to mock store or to pass the data, which is expanding the purpose of encodeData. The same goes for decodeData.

For decodeData, would the Get fail here https://github.com/ethersphere/go-ethereum/blob/master/swarm/storage/dbstore.go#L482, as the data will not be valid from the database, the decodeData would retrieve the data from mock store later in that function?

My two concerns here are, is additional *if* accepted in every Put and Get function call to check if mock store is provided and the issue with validation of empty data on https://github.com/ethersphere/go-ethereum/blob/master/swarm/storage/dbstore.go#L482, if decodeData handles data retrieval from mock store.